___

<a href='http://www.pieriandata.com'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Copyright by Pierian Data Inc.</em></center>
<center><em>For more information, visit us at <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

# Linear Regression Project Exercise 

Now that we have learned about feature engineering, cross validation, and grid search, let's test all your new skills with a project exercise in Machine Learning. This exercise will have a more guided approach, later on the ML projects will begin to be more open-ended. We'll start off with using the final version of the Ames Housing dataset we worked on through the feature engineering section of the course. Your goal will be to create a Linear Regression Model, train it on the data with the optimal parameters using a grid search, and then evaluate the model's capabilities on a test set.

---
---
---
## Complete the tasks in bold

**TASK: Run the cells under the Imports and Data section to make sure you have imported the correct general libraries as well as the correct datasets. Later on you may need to run further imports from scikit-learn.**

### Imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### Data

In [2]:
df = pd.read_csv("../DATA/AMES_Final_DF.csv")

In [3]:
df.head()

Lot Frontage  Lot Area  Overall Qual  Overall Cond  Year Built  \
0         141.0     31770             6             5        1960   
1          80.0     11622             5             6        1961   
2          81.0     14267             6             6        1958   
3          93.0     11160             7             5        1968   
4          74.0     13830             5             5        1997   

   Year Remod/Add  Mas Vnr Area  BsmtFin SF 1  BsmtFin SF 2  Bsmt Unf SF  ...  \
0            1960         112.0         639.0           0.0        441.0  ...   
1            1961           0.0         468.0         144.0        270.0  ...   
2            1958         108.0         923.0           0.0        406.0  ...   
3            1968           0.0        1065.0           0.0       1045.0  ...   
4            1998           0.0         791.0           0.0        137.0  ...   

   Sale Type_ConLw  Sale Type_New  Sale Type_Oth  Sale Type_VWD  \
0                0              0              0              0   
1                0              0              0              0   
2                0              0              0              0   
3                0              0              0              0   
4                0              0              0              0   

   Sale Type_WD   Sale Condition_AdjLand  Sale Condition_Alloca  \
0              1                       0                      0   
1              1                       0                      0   
2              1                       0                      0   
3              1                       0                      0   
4              1                       0                      0   

   Sale Condition_Family  Sale Condition_Normal  Sale Condition_Partial  
0                      0                      1                       0  
1                      0                      1                       0  
2                      0                      1                       0  
3                      0                      1                       0  
4                      0                      1                       0  

[5 rows x 274 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Columns: 274 entries, Lot Frontage to Sale Condition_Partial
dtypes: float64(11), int64(263)
memory usage: 6.1 MB


**TASK: The label we are trying to predict is the SalePrice column. Separate out the data into X features and y labels**

In [5]:
X=df.drop('SalePrice',axis=1)
y=df['SalePrice']

In [6]:
X.head()

Lot Frontage  Lot Area  Overall Qual  Overall Cond  Year Built  \
0         141.0     31770             6             5        1960   
1          80.0     11622             5             6        1961   
2          81.0     14267             6             6        1958   
3          93.0     11160             7             5        1968   
4          74.0     13830             5             5        1997   

   Year Remod/Add  Mas Vnr Area  BsmtFin SF 1  BsmtFin SF 2  Bsmt Unf SF  ...  \
0            1960         112.0         639.0           0.0        441.0  ...   
1            1961           0.0         468.0         144.0        270.0  ...   
2            1958         108.0         923.0           0.0        406.0  ...   
3            1968           0.0        1065.0           0.0       1045.0  ...   
4            1998           0.0         791.0           0.0        137.0  ...   

   Sale Type_ConLw  Sale Type_New  Sale Type_Oth  Sale Type_VWD  \
0                0              0              0              0   
1                0              0              0              0   
2                0              0              0              0   
3                0              0              0              0   
4                0              0              0              0   

   Sale Type_WD   Sale Condition_AdjLand  Sale Condition_Alloca  \
0              1                       0                      0   
1              1                       0                      0   
2              1                       0                      0   
3              1                       0                      0   
4              1                       0                      0   

   Sale Condition_Family  Sale Condition_Normal  Sale Condition_Partial  
0                      0                      1                       0  
1                      0                      1                       0  
2                      0                      1                       0  
3                      0                      1                       0  
4                      0                      1                       0  

[5 rows x 273 columns]

In [7]:
y.head()

0    215000
1    105000
2    172000
3    244000
4    189900
Name: SalePrice, dtype: int64

**TASK: Use scikit-learn to split up X and y into a training set and test set. Since we will later be using a Grid Search strategy, set your test proportion to 10%. To get the same data split as the solutions notebook, you can specify random_state = 101**

In [8]:
from sklearn.model_selection import train_test_split

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=101)

**TASK: The dataset features has a variety of scales and units. For optimal regression performance, scale the X features. Take carefuly note of what to use for .fit() vs what to use for .transform()**

In [10]:
from sklearn.preprocessing import StandardScaler

In [11]:
scaler= StandardScaler()

In [12]:
scaled_X_train= scaler.fit_transform(X_train)
scaled_X_test= scaler.transform(X_test)

**TASK: We will use an Elastic Net model. Create an instance of default ElasticNet model with scikit-learn**

In [13]:
from sklearn.linear_model import ElasticNet

In [14]:
model= ElasticNet()

**TASK:  The Elastic Net model has two main parameters, alpha and the L1 ratio. Create a dictionary parameter grid of values for the ElasticNet. Feel free to play around with these values, keep in mind, you may not match up exactly with the solution choices**

ElasticNet(
    alpha=1.0,
    *,
    l1_ratio=0.5,
    fit_intercept=True,
    normalize='deprecated',
    precompute=False,
    max_iter=1000,
    copy_X=True,
    tol=0.0001,
    warm_start=False,
    positive=False,
    random_state=None,
    selection='cyclic',
)
Docstring:     
Linear regression with combined L1 and L2 priors as regularizer.
alpha : float, default=1.0
    Constant that multiplies the penalty terms. Defaults to 1.0.
    
l1_ratio : float, default=0.5
    The ElasticNet mixing parameter, with ``0 <= l1_ratio <= 1``. 
    For ``l1_ratio = 0`` the penalty is an L2 penalty.
    ``For l1_ratio = 1`` it is an L1 penalty.  
    For ``0 < l1_ratio < 1``, the penalty is a combination of L1 and L2.

In [17]:
#create dict for diff values of alpha & l1 ration
param_grid= {'alpha': [0.1,1,5,10,50,100],
            'l1_ratio': [0.1,0.5,0.7,0.9,0.95,0.99,1]}

**TASK: Using scikit-learn create a GridSearchCV object and run a grid search for the best parameters for your model based on your scaled training data. [In case you are curious about the warnings you may recieve for certain parameter combinations](https://stackoverflow.com/questions/20681864/lasso-on-sklearn-does-not-converge)**

In [18]:
from sklearn.model_selection import GridSearchCV

In [19]:
grid_model=GridSearchCV(estimator=model,param_grid=param_grid,cv=5,verbose=1)

In [20]:
grid_model.fit(scaled_X_train,y_train)

Fitting 5 folds for each of 42 candidates, totalling 210 fits


C:\Users\manis\anaconda3\envs\new_env\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.394e+11, tolerance: 1.355e+09
  model = cd_fast.enet_coordinate_descent(
C:\Users\manis\anaconda3\envs\new_env\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.654e+11, tolerance: 1.308e+09
  model = cd_fast.enet_coordinate_descent(
C:\Users\manis\anaconda3\envs\new_env\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:647: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularis

GridSearchCV(cv=5, estimator=ElasticNet(),
             param_grid={'alpha': [0.1, 1, 5, 10, 50, 100],
                         'l1_ratio': [0.1, 0.5, 0.7, 0.9, 0.95, 0.99, 1]},
             verbose=1)

**TASK: Display the best combination of parameters for your model**

In [21]:
grid_model.best_params_

{'alpha': 100, 'l1_ratio': 1}

**TASK: Evaluate your model's performance on the unseen 10% scaled test set.

In [22]:
y_prediction = grid_model.predict(scaled_X_test)

In [23]:
y_prediction

array([115030.0108828 , 330700.15372654, 201207.12731429, 196378.72400031,
       102219.74319082, 149001.41833178, 130072.86469072, 344190.99699698,
        86317.81211648, 214611.83988343, 280702.74310865, 133181.46547467,
       313228.76346267, 110025.52409208, 162260.28149297, 235688.91560529,
       212164.9071257 , 139729.23537012, 166151.3842688 , 138201.59345812,
       163063.02559551, 217573.52020481, 252731.55626605, 225696.62317213,
       143089.79702881, 180148.28076638, 126670.91277983, 259872.59348454,
       186330.54115183, 217235.54894129,  72441.18646171, 117379.48599897,
       153390.9769354 , 191146.92308306, 247299.30063386,  90128.32765601,
        91978.63317599, 135154.72698894, 212033.10745544, 211636.19384069,
       165084.76510891, 269665.98189404, 196236.07683552, 157678.75420519,
       169504.22735377, 291698.85706455, 108640.38433264, 106942.01149915,
       407661.01966128, 238562.39217868, 143419.85167273, 199431.46443486,
       228893.12046043, 3

In [24]:
from sklearn.metrics import mean_absolute_error,mean_squared_error

In [26]:
mean_absolute_error(y_test,y_prediction)

14195.354900562166

In [27]:
mean_squared_error(y_test,y_prediction)

422652274.49501973

In [28]:
np.sqrt(mean_squared_error(y_test,y_prediction))

20558.508566893168

In [29]:
#lets compare to mean of our label i.e.,saleprice (average sale price of a house)
np.mean(df['SalePrice'])

180815.53743589742

It seems our models is off by 14195(14k) on MAE & 20558(20k) on RMSE.If its too much or not ,totally depends on house pricing depending on different cities & their profitibility.

## Great work!

----